### DISCUSSION TITLE
Performance Difference

### DISCUSSION THREAD

**Discussion by ronhombre** [ronhombre/KeccakKotlin#1] [Show and tell]
This is an open letter addressed to @05nelsonm.

Hello!

I was very satisfied with your keccak and sha3 libraries until I found out through the IntelliJ Profiler that they were the slowest part of my Kyber implementation. Thus, I have written my own. I would like to know your thoughts and look forward to sharing ideas regarding our implementations.

Before(Your implementation):
![image](https://github.com/ronhombre/KeccakKotlin/assets/118486316/ccbe5b9f-f967-47ac-8153-08e58694a3ab)
After(My implementation):
![image](https://github.com/ronhombre/KeccakKotlin/assets/118486316/3901ecd3-4920-41ac-928e-ee2d859dd305)

Best Regards,
Ron Lauren Hombre

**Comment by 05nelsonm** [comment]
Seems like the [KotlinCrypto/sponges/KeccakP](https://github.com/KotlinCrypto/sponges/blob/master/library/keccak/src/commonMain/kotlin/org/kotlincrypto/sponges/keccak/KeccakP.kt) and [KotlinCrypto/sponges/State.addData](https://github.com/KotlinCrypto/sponges/blob/d7c5eb53d5717b3f9b607d337d7c4b3e5d644f3e/library/keccak/src/commonMain/kotlin/org/kotlincrypto/sponges/keccak/State.kt#L57) functions could benefit greatly from some optimizations.

I don't have much time atm to dig into it, but you can open an issue ticket [HERE](https://github.com/KotlinCrypto/sponges/issues/new) if you'd like.

**Reply by 05nelsonm** [reply]
Did some optimizing in `sponges` and `digest` which resulted in some huge performance improvements.

See benchmark comparisons at https://github.com/KotlinCrypto/hash/pull/70#issuecomment-2554683182

**Reply by ronhombre** [reply]
> Did some optimizing in `sponges` and `digest` which resulted in some huge performance improvements.
> 
> See benchmark comparisons at https://github.com/KotlinCrypto/hash/pull/70#issuecomment-2554683182

Sounds great! I will compare it with my keccak library when I have the time

**Reply by 05nelsonm** [reply]
Was still slow. Figured it out. `kotlincrypto.sponges:keccak` implementation was utilizing generics which was causing Kotlin to compile with `java.lang.Long` and other boxed numbers instead of primitives. With `0.5.5` release of `hash`, I threw BouncyCastle, KotlinCrypto, and this library on a benchmark to see comparisons:

```
jvm summary:
Benchmark                       Mode  Cnt     Score     Error  Units
BCSHA3_256Benchmark.digest      avgt    5   533.168 ± 113.096  ns/op
BCSHAKE128Benchmark.digest      avgt    5   541.318 ±   7.140  ns/op
KCSHA3_256Benchmark.digest      avgt    5   897.816 ±   2.260  ns/op
KCSHAKE128Benchmark.digest      avgt    5   948.368 ±   2.530  ns/op
KeccakSHA3_256Benchmark.digest  avgt    5  2046.828 ± 203.918  ns/op
KeccakSHAKE128Benchmark.digest  avgt    5  2147.210 ± 182.244  ns/op
```
